File: KAFKA_KOPANDA/src/main/kotlin/com/sleekydz86/kopanda/domain/entities/Alert.kt
package com.sleekydz86.kopanda.domain.entities

import com.sleekydz86.kopanda.domain.valueobjects.common.IssueSeverity
import com.sleekydz86.kopanda.domain.valueobjects.common.IssueType
import com.sleekydz86.kopanda.shared.domain.AggregateRoot
import java.time.LocalDateTime

class Alert(
    val type: IssueType,
    val severity: IssueSeverity,
    val title: String,
    val message: String,
    val connectionId: String?,
    val topicName: String?,
    val partitionNumber: Int?,
    val timestamp: LocalDateTime = LocalDateTime.now(),
    var isAcknowledged: Boolean = false,
    var acknowledgedAt: LocalDateTime? = null,
    var acknowledgedBy: String? = null
) : AggregateRoot() {

    val id: String = java.util.UUID.randomUUID().toString()

    fun acknowledge(acknowledgedBy: String) {
        this.isAcknowledged = true
        this.acknowledgedAt = LocalDateTime.now()
        this.acknowledgedBy = acknowledgedBy
    }

    fun isActive(): Boolean = !isAcknowledged

    fun toDto() = com.sleekydz86.kopanda.application.dto.response.AlertDto(
        id = id,
        type = type,
        severity = severity,
        title = title,
        message = message,
        connectionId = connectionId,
        topicName = topicName,
        partitionNumber = partitionNumber,
        timestamp = timestamp,
        isAcknowledged = isAcknowledged,
        acknowledgedAt = acknowledgedAt,
        acknowledgedBy = acknowledgedBy
    )

    companion object {
        fun create(
            type: IssueType,
            severity: IssueSeverity,
            title: String,
            message: String,
            connectionId: String? = null,
            topicName: String? = null,
            partitionNumber: Int? = null
        ): Alert {
            return Alert(
                type = type,
                severity = severity,
                title = title,
                message = message,
                connectionId = connectionId,
                topicName = topicName,
                partitionNumber = partitionNumber
            )
        }
    }
}

File: KAFKA_PANDA_FRONT/src/types/alert.ts
export interface AlertDto {
  id: string
  type: IssueType
  severity: IssueSeverity
  title: string
  message: string
  connectionId?: string
  topicName?: string
  partitionNumber?: number
  timestamp: string
  isAcknowledged: boolean
  acknowledgedAt?: string
  acknowledgedBy?: string
}

export enum IssueType {
  UNDER_REPLICATED = 'UNDER_REPLICATED',
  OFFLINE_PARTITION = 'OFFLINE_PARTITION',
  LEADER_NOT_AVAILABLE = 'LEADER_NOT_AVAILABLE',
  REPLICATION_FACTOR_MISMATCH = 'REPLICATION_FACTOR_MISMATCH',
  CONSUMER_LAG = 'CONSUMER_LAG',
  CONNECTION_ERROR = 'CONNECTION_ERROR'
}

export enum IssueSeverity {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL'
}

export interface CreateTestAlertRequest {
  connectionId: string
}

export interface AcknowledgeAlertRequest {
  alertId: string
  acknowledgedBy: string
}

export interface ClearAllAlertsRequest {
  connectionId?: string
}
